storage: rewrite TestStoreRangeRebalance

petermattis opened the pull request:
Rather than the somewhat complicated rebalancing scenario, use a simple
scenario that we perform up-replication of range 1 from 1 to 3 nodes. We
check that this up-replication is performed using preemptive
snapshots. The more complicated scenario was very fragile, frequently
being broken by innocuous changes.

Fixes #10497
Fixes #10193
Fixes #10156
Fixes #9395

<!-- Reviewable:start -->

---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/10515)
<!-- Reviewable:end -->

petermattis committed:
storage: remove TestStoreRangeRebalance

Add the check that preemptive snapshots are being used to
TestStoreRangeUpReplicate. Add TestReplicateQueueRebalance for testing
that basic rebalancing is working.

Fixes #10193
Fixes #10156
Fixes #9395

petermattis commented:
@tamird Note that #10501 goes away with this change, but I think we should track down why that failure was occurring.

tamird commented:
Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[pkg/storage/client_raft_test.go, line 2040 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10515#-KW-bNOQAxq2eEfH3bVP:-KW-bNORPE_sDHDMJZrV:b-hjuznc) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e1e31c08fc3f858f685036188f623d77121df9b/pkg/storage/client_raft_test.go#L2040)):*
> ```Go
> }
> 
> // TestStoreRangeRebalance verifies that the replication queue will take
> ```

s/ take// and re-wrap this guy

---

*[pkg/storage/client_raft_test.go, line 2042 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10515#-KW-bd3a2wLismNv6nRy:-KW-bd3a2wLismNv6nRz:b-gp3gq5) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e1e31c08fc3f858f685036188f623d77121df9b/pkg/storage/client_raft_test.go#L2042)):*
> ```Go
> // TestStoreRangeRebalance verifies that the replication queue will take
> // up-replicate a range when possible.
> func TestStoreRangeRebalance(t *testing.T) {
> ```

this test name doesn't make sense. The test is now the equivalent of `testcluster.TestBasicAutoReplication`?

---

*[pkg/storage/client_raft_test.go, line 2045 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10515#-KW-b_V72S5BZh_71ZuR:-KW-b_V72S5BZh_71ZuS:b5av6wn) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e1e31c08fc3f858f685036188f623d77121df9b/pkg/storage/client_raft_test.go#L2045)):*
> ```Go
> 	defer leaktest.AfterTest(t)()
> 
> 	// Start multiTestContext with replica rebalancing enabled.
> ```

this is nonsense now, right? no rebalancing will take place in this test no matter what options are passed.

---

*[pkg/storage/client_raft_test.go, line 2058 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10515#-KW-c-Da0FSrTGhT6paW:-KW-c-Da0FSrTGhT6paX:b-vpyxn8) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e1e31c08fc3f858f685036188f623d77121df9b/pkg/storage/client_raft_test.go#L2058)):*
> ```Go
> 
> 		// Exit when all stores have a single replica.
> 		const expected = 3
> ```

duplicates the argument to `mtc.Start`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10515)*
<!-- Sent from Reviewable.io -->

tamird commented:
BTW, I don't think it's safe to say that this fixes #10193 and #10497 - those failures produce nonsense error messages like

> client_raft_test.go:2093: range 1: replica {2 2 2} not lease holder; node_id:2 store_id:2 replica_id:2  is

which definitely are worth tracking down.

---

Review status: 2 of 3 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10515#-:-KW-f9n__7twxCnEr24U:b-6e666o)*
<!-- Sent from Reviewable.io -->

petermattis commented:
> which definitely are worth tracking down.

Those are due to periodic gossiping of the first range for which we're also transferring the lease in unexpected ways. We can't disable that periodic gossiping in the previous incarnation of this test or the test doesn't make progress.

---

Review status: 2 of 3 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[pkg/storage/client_raft_test.go, line 2040 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10515#-KW-bNOQAxq2eEfH3bVP:-KW-fIgAiwvLdFNTezWH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e1e31c08fc3f858f685036188f623d77121df9b/pkg/storage/client_raft_test.go#L2040)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/ take// and re-wrap this guy
</details>
Done.

---

*[pkg/storage/client_raft_test.go, line 2042 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10515#-KW-bd3a2wLismNv6nRy:-KW-fIzBEQ-Nt5B56DtI:b-f3fuye) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e1e31c08fc3f858f685036188f623d77121df9b/pkg/storage/client_raft_test.go#L2042)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this test name doesn't make sense. The test is now the equivalent of `testcluster.TestBasicAutoReplication`?
</details>
More or less. It also verifies that we're using preemptive snapshots. Not sure if that is worth a separate test here.

---

*[pkg/storage/client_raft_test.go, line 2045 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10515#-KW-b_V72S5BZh_71ZuR:-KW-f_zMAz5f5_KE5GP2:b-ncu9h4) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e1e31c08fc3f858f685036188f623d77121df9b/pkg/storage/client_raft_test.go#L2045)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this is nonsense now, right? no rebalancing will take place in this test no matter what options are passed.
</details>
Removed and simplified.

---

*[pkg/storage/client_raft_test.go, line 2058 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10515#-KW-c-Da0FSrTGhT6paW:-KW-fbblTWLZopQoeoAq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e1e31c08fc3f858f685036188f623d77121df9b/pkg/storage/client_raft_test.go#L2058)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> duplicates the argument to `mtc.Start`
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10515#-:-KW-fpbN8CQyAtCIV7LO:b-miwp09)*
<!-- Sent from Reviewable.io -->

tamird commented:
As mentioned in person, the troubling thing about those error messages is that they are saying "replica X is not the leaseholder, X is", which is obviously nonsense.

---

Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/storage/client_raft_test.go, line 2042 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10515#-KW-bd3a2wLismNv6nRy:-KW-gkiKGKUmnpQMyPNu:b-xk7iq5) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e1e31c08fc3f858f685036188f623d77121df9b/pkg/storage/client_raft_test.go#L2042)):*
> It also verifies that we're using preemptive snapshots.

Ah, right, didn't notice that there was still more stuff below the retry loop.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10515#-:-KW-guNUImxwh0Uoiici:bkquypx)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 2 of 3 files at r1, 1 of 1 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/storage/client_raft_test.go, line 2042 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10515#-KW1t0nTIGHdRNggajke:-KW1t0nTIGHdRNggajkf:bfal5p) ([raw file](https://github.com/cockroachdb/cockroach/blob/efd4ada808939f1c7498238feb38544ad223838a/pkg/storage/client_raft_test.go#L2042)):*
> ```Go
> // TestStoreRangeUpreplicate verifies that the replication queue will
> // up-replicate a range when possible.
> func TestStoreRangeUpreplicate(t *testing.T) {
> ```

We already have the same test in TestStoreRangeUpReplicate (with a capital R), so if there's nothing to salvage from TestStoreRangeRebalance we should just delete it. But I think the old test is worth fixing and keeping - upreplication and rebalancing are separate code paths, and I don't think we have much other testing of the rebalancing path. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10515)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 3 of 5 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/storage/client_raft_test.go, line 2042 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10515#-KW1t0nTIGHdRNggajke:-KW3FoXMZK6zeEElGgAC:b-cl20v1) ([raw file](https://github.com/cockroachdb/cockroach/blob/efd4ada808939f1c7498238feb38544ad223838a/pkg/storage/client_raft_test.go#L2042)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> We already have the same test in TestStoreRangeUpReplicate (with a capital R), so if there's nothing to salvage from TestStoreRangeRebalance we should just delete it. But I think the old test is worth fixing and keeping - upreplication and rebalancing are separate code paths, and I don't think we have much other testing of the rebalancing path. 
</details>
Ah, didn't realize we already had `TestStoreRangeUpReplicate`. I've modified that test to verify we're performing preemptive snapshots during up-replication. I've removed `TestStoreRangeRebalance` completely and add a new `TestReplicateQueueRebalance` that uses a `TestCluster` to verify that basic rebalancing is occurring. I've also nuked `multiTestContext.transferLease()`.

PTAL.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10515)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r3, 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/storage/client_raft_test.go, line 2042 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10515#-KW1t0nTIGHdRNggajke:-KW3X3TTxZri2Jo73m5I:bnjzbbv) ([raw file](https://github.com/cockroachdb/cockroach/blob/efd4ada808939f1c7498238feb38544ad223838a/pkg/storage/client_raft_test.go#L2042)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Ah, didn't realize we already had `TestStoreRangeUpReplicate`. I've modified that test to verify we're performing preemptive snapshots during up-replication. I've removed `TestStoreRangeRebalance` completely and add a new `TestReplicateQueueRebalance` that uses a `TestCluster` to verify that basic rebalancing is occurring. I've also nuked `multiTestContext.transferLease()`.
> 
> PTAL.
</details>
:tada:

---

*[pkg/storage/replicate_queue_test.go, line 40 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10515#-KW3X8JjeAIZdD38he22:-KW3X8JjeAIZdD38he23:bbx2o56) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b43b62138fb62017be574ee3eb8064daa563422/pkg/storage/replicate_queue_test.go#L40)):*
> ```Go
> 	// Set the gossip stores interval lower to speed up rebalancing. With the
> 	// default of 5s we have to wait ~5s for the rebalancing to start.
> 	_ = os.Setenv("COCKROACH_GOSSIP_STORES_INTERVAL", "100ms")
> ```

can we check this error, please? I actually don't know under what circumstance this can return a non-nil error, but this kind of thing makes me nervous and isn't worth the two line savings. ditto below.

---

*[pkg/storage/replicate_queue_test.go, line 60 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10515#-KW3XhspROIDW7tGA1aX:-KW3Xhsq4th5DeY57NHI:b-m9vfml) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b43b62138fb62017be574ee3eb8064daa563422/pkg/storage/replicate_queue_test.go#L60)):*
> ```Go
> 		for {
> 			if _, _, err := tc.SplitRange(splitKey); err != nil {
> 				// The split can conflict with rebalancing. Log the error and keep
> ```

can we do a more targeted error check here?

---

*[pkg/storage/replicate_queue_test.go, line 85 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10515#-KW3Xvk2kYQy6I75gYm0:-KW3Xvk2kYQy6I75gYm1:blqn0kt) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b43b62138fb62017be574ee3eb8064daa563422/pkg/storage/replicate_queue_test.go#L85)):*
> ```Go
> 		counts := countReplicas()
> 		for _, c := range counts {
> 			if c == 0 {
> ```

this is not really checking that things are balanced at all, is it? it's only checking that every node has at least one replica, which is contrary to the detailed comment way above that lists the expected numbers of ranges, etc.

Can we do a "real" check of balancedness here? Also if we do, we should remove magic numbers from that comment since those numbers can and will rot.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10515)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 4 of 5 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replicate_queue_test.go, line 40 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10515#-KW3X8JjeAIZdD38he22:-KW3_tbunO_xZXihrmc8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b43b62138fb62017be574ee3eb8064daa563422/pkg/storage/replicate_queue_test.go#L40)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> can we check this error, please? I actually don't know under what circumstance this can return a non-nil error, but this kind of thing makes me nervous and isn't worth the two line savings. ditto below.
</details>
Done.

---

*[pkg/storage/replicate_queue_test.go, line 60 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10515#-KW3XhspROIDW7tGA1aX:-KW3aAPserk4q2FuFv2X:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b43b62138fb62017be574ee3eb8064daa563422/pkg/storage/replicate_queue_test.go#L60)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> can we do a more targeted error check here?
</details>
Done.

---

*[pkg/storage/replicate_queue_test.go, line 85 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10515#-KW3Xvk2kYQy6I75gYm0:-KW3ZjRNbNoL2N3ygqtF:br7hro9) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b43b62138fb62017be574ee3eb8064daa563422/pkg/storage/replicate_queue_test.go#L85)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this is not really checking that things are balanced at all, is it? it's only checking that every node has at least one replica, which is contrary to the detailed comment way above that lists the expected numbers of ranges, etc.
> 
> Can we do a "real" check of balancedness here? Also if we do, we should remove magic numbers from that comment since those numbers can and will rot.
</details>
Testing "balancedness" is extremely difficult to do robustly (if you feel otherwise, do a follow-up PR where you improve this test). I think of this test as a basic sanity check that rebalancing is being performed, not that the heuristics are working correctly. To test the heuristics, we should have higher level tests powered by infrastructure like `allocsim`. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10515)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 4 of 5 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replicate_queue_test.go, line 51 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10515#-KW3bttNdTdgozfWvnSe:-KW3bttNdTdgozfWvnSf:b9970qp) ([raw file](https://github.com/cockroachdb/cockroach/blob/de3434de2aa916df005ccea8a860e2b80a8abdda/pkg/storage/replicate_queue_test.go#L51)):*
> ```Go
> 
> 	// TODO(peter): Bump this to 10 nodes. Doing so is flaky (we encounter a
> 	// situation where 1 node never receives a replica). Need to figure out why.
> ```

I'm investigating this before merging. The allocator thinks that everything is balanced while `countReplicas` does not. That discrepancy is disturbing. Also disturbing is that the discrepancy persists given that we're supposed to be gossiping the store capacity every 100ms. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10515)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replicate_queue_test.go, line 85 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10515#-KW3Xvk2kYQy6I75gYm0:-KW3dk98QmBs01AaSNl6:b-1o7ilr) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b43b62138fb62017be574ee3eb8064daa563422/pkg/storage/replicate_queue_test.go#L85)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Testing "balancedness" is extremely difficult to do robustly (if you feel otherwise, do a follow-up PR where you improve this test). I think of this test as a basic sanity check that rebalancing is being performed, not that the heuristics are working correctly. To test the heuristics, we should have higher level tests powered by infrastructure like `allocsim`. 
</details>
That's fine, but then let's not lie about it in the top level comment that implies we're going to actually check range counts per store.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10515)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 4 of 5 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replicate_queue_test.go, line 51 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10515#-KW3bttNdTdgozfWvnSe:-KW3gXvHba053VvGWH7M:bz9rk17) ([raw file](https://github.com/cockroachdb/cockroach/blob/de3434de2aa916df005ccea8a860e2b80a8abdda/pkg/storage/replicate_queue_test.go#L51)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> I'm investigating this before merging. The allocator thinks that everything is balanced while `countReplicas` does not. That discrepancy is disturbing. Also disturbing is that the discrepancy persists given that we're supposed to be gossiping the store capacity every 100ms. 
</details>
Figured this out. Without lease rebalancing we can get into a situation where 1 store has 0 replicas. I could increase the number of replicas in the test, but seems better to wait for lease rebalancing to land.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10515)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replicate_queue_test.go, line 85 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10515#-KW3Xvk2kYQy6I75gYm0:-KW3hQYlnl065RZnn241:b-t2gylu) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b43b62138fb62017be574ee3eb8064daa563422/pkg/storage/replicate_queue_test.go#L85)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> That's fine, but then let's not lie about it in the top level comment that implies we're going to actually check range counts per store.
</details>
LGTM modulo this comment.

---

*[pkg/storage/replicate_queue_test.go, line 51 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10515#-KW3hNM1RXJlf5Ysr-i1:-KW3hNM1RXJlf5Ysr-i2:bbu4k8l) ([raw file](https://github.com/cockroachdb/cockroach/blob/d71fa95c99d8497dc151cfae00e6c6b12db87b26/pkg/storage/replicate_queue_test.go#L51)):*
> we can achieve the store 1 can hold on to too many

broken sentence

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10515)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 4 of 5 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/storage/replicate_queue_test.go, line 85 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10515#-KW3Xvk2kYQy6I75gYm0:-KW3hvaXEROy5uTJeAVK:b7sim0a) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b43b62138fb62017be574ee3eb8064daa563422/pkg/storage/replicate_queue_test.go#L85)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> LGTM modulo this comment.
</details>
TODO added and expanded the comment above.

---

*[pkg/storage/replicate_queue_test.go, line 51 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10515#-KW3hNM1RXJlf5Ysr-i1:-KW3hdvagbwnq4Wo9LGg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d71fa95c99d8497dc151cfae00e6c6b12db87b26/pkg/storage/replicate_queue_test.go#L51)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> > we can achieve the store 1 can hold on to too many
> 
> broken sentence
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10515)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10515)*
<!-- Sent from Reviewable.io -->
